Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigation.swift

```diff
@@ -0,0 +1,520 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018-2021 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import UIKit
+
+
+public protocol BPKHorizontalNavigationSize {
+    static var fontStyle: BPKFontStyle { get }
+
+    static var verticalSpacing: CGFloat { get }
+}
+
+public enum BPKHorizontalNavigationSizeDefault: BPKHorizontalNavigationSize {
+    public static var verticalSpacing: CGFloat { BPKSpacingLg }
+
+    public static var fontStyle: BPKFontStyle { .textBaseEmphasized }
+}
+
+public enum BPKHorizontalNavigationSizeSmall: BPKHorizontalNavigationSize {
+    public static var verticalSpacing: CGFloat { BPKSpacingBase }
+
+    public static var fontStyle: BPKFontStyle { .textSmEmphasized }
+}
+
+extension BPKHorizontalNavigationSize {
+    static var horizontalSpacing: CGFloat { BPKSpacingBase }
+}
+
+public protocol BPKHorizontalNavigationDelegate: AnyObject {
+    func horizontalNavigation<Size>(_ horizontalNavigation: BPKHorizontalNavigation<Size>, didSelectItemAtIndex index: NSInteger, withTag tag: NSInteger)
+}
+
+@objc
+public enum BPKHorizontalNavigationAppearance: UInt {
+    case normal = 0
+    case alternate = 1
+}
+
+public protocol BPKHorizontalNavigationItem: UIControl {
+    var selectedColor: UIColor? { get set }
+
+    var appearance: BPKHorizontalNavigationAppearance { get set }
+}
+
+public protocol BPKHorizontalNavigationOption {
```

Comment:
This seems to be an ItemProvider instead of an Option instance.
Maybe a renderer.
wdyt about this approach?

```swift
public protocol ViewItem: UIControl { }

public protocol ViewRenderer {
   func render() -> ViewItem
}

// Examples
protocol SelectableViewItem: ViewItem {
   var selectedColor: UIColor { get }
}

class CellViewItem: SelectableViewItem {
    var selectedColor: UIColor = .red
}

public class SelectableViewRenderer: ViewRenderer {
   func renderer() -> ViewItem {
      return CellViewItem()
   }
}

class HorizontalNavigationItemWithBackground: ViewItem {}

class HorizontalNavigationOptionWithBackgroundRenderer<Size: BPKHNSize>: ViewRenderer {
   func render() -> ViewItem {
      return HorizontalNavigationItemWithBackground()
   }
}

class Container: ViewItem {}

class ListComposerRenderer: ViewRenderer {
   let renderers: [ViewRenderer] = [...]
   func render() -> ViewItem {
      let container = Container()
      renderers
            .map { $0.render() }
            .forEach(container.addSubview)
     return container
   }
}

// Usage
class HorizontalNavigation {
   init(renderer: ViewRenderer) {
      let renderedControl = renderer.render()
   }
}
```

Reply:
we can discuss further about this later

Reply:
I think this providers slightly more flexibility than desirable 🤔 

The thinking here is that `Option` represents the combination of model and the knowledge about how to create a view to render that model. I guess the best example of this is `UITabBarItem` which is kind of like our `Option` here and then there's some other view that we don't have access to that is actually rendered. The `Option` is exposed to the user whereas the view is hidden from them, you can provide custom options by defining your own option and item.

We want all potential items that we render to support both `selectedColor` and `appearance` which is why that's part of the `BPKHorizontalNavigationItem` protocol atm.